fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -776,175 +886,214 @@ ACTOR Future<Void> releaseResolvingAfter(ProxyCommitData* self, Future<Void> rel
+			trs[t].transaction.read_conflict_ranges.size() +
+			trs[t].transaction.write_conflict_ranges.size()
+		);
 		//TraceEvent("MPTransactionDump", self->dbgid).detail("Snapshot", trs[t].transaction.read_snapshot);
 		//for(auto& m : trs[t].transaction.mutations)
-		maxTransactionBytes = std::max<int64_t>(maxTransactionBytes, trs[t].transaction.expectedSize());
+		context.maxTransactionBytes = std::max<int64_t>(
+			context.maxTransactionBytes, trs[t].transaction.expectedSize()
+		);
 		//	TraceEvent("MPTransactionsDump", self->dbgid).detail("Mutation", m.toString());
 	}
-	self->stats.conflictRanges += conflictRangeCount;
+	pProxyCommitData->stats.conflictRanges += conflictRangeCount;
 
-	for (int r = 1; r<self->resolvers.size(); r++)
-		ASSERT(requests.requests[r].txnStateTransactions.size() == requests.requests[0].txnStateTransactions.size());
+	for (int r = 1; r < pProxyCommitData->resolvers.size(); r++)
+		ASSERT(requests.requests[r].txnStateTransactions.size() ==
+			requests.requests[0].txnStateTransactions.size());
 
-	// Sending these requests is the fuzzy border between phase 1 and phase 2; it could conceivably overlap with resolution processing but is still using CPU
-	self->stats.txnCommitResolving += trs.size();
-	vector< Future<ResolveTransactionBatchReply> > replies;
-	for (int r = 0; r<self->resolvers.size(); r++) {
-		requests.requests[r].debugID = debugID;
-		replies.push_back(brokenPromiseToNever(self->resolvers[r].resolve.getReply(requests.requests[r], TaskPriority::ProxyResolverReply)));
+	pProxyCommitData->stats.txnCommitResolving += trs.size();
+	std::vector<Future<ResolveTransactionBatchReply>> replies;
+	for (int r = 0; r < pProxyCommitData->resolvers.size(); r++) {
+		requests.requests[r].debugID = context.debugID;
+		replies.push_back(brokenPromiseToNever(
+			pProxyCommitData->resolvers[r].resolve.getReply(
+				requests.requests[r], TaskPriority::ProxyResolverReply)));
 	}
 
-	state vector<vector<int>> transactionResolverMap = std::move( requests.transactionResolverMap );
-	state std::vector<std::vector<std::vector<int>>> txReadConflictRangeIndexMap =
-	    std::move(requests.txReadConflictRangeIndexMap); // used to report conflicting keys
-	state Future<Void> releaseFuture = releaseResolvingAfter(self, releaseDelay, localBatchNumber);
+	context.transactionResolverMap.swap(requests.transactionResolverMap);
+	// Used to report conflicting keys
+	context.txReadConflictRangeIndexMap.swap(requests.txReadConflictRangeIndexMap);
+	context.releaseFuture = releaseResolvingAfter(pProxyCommitData, context.releaseDelay, context.localBatchNumber);
 
-	/////// Phase 2: Resolution (waiting on the network; pipelined)
-	state vector<ResolveTransactionBatchReply> resolution = wait( getAll(replies) );
+	// Wait for the final resolution
+	std::vector<ResolveTransactionBatchReply> resolutionResp = wait(getAll(replies));
+	context.resolution.swap(*const_cast<std::vector<ResolveTransactionBatchReply>*>(&resolutionResp));
```

Comment:
I don't think cast is needed here.
```suggestion
	context.resolution.swap(resolutionResp);
```

Reply:
This is a really interesting issue. So at first I also thought a simple swap should work. However this is not true.

 * The `swap` statement will be transpiled into a `a_call_back_file` function, the function would have both `const vector&` version and `vector&&` version.
 * The generated `a_call_back_file` would use `ActorCallback` class rather than `ActorSingleCallback`.
 * `ActorCallback` class has only `fire(ValueType const&)` method, without `fire(ValueType&&)` type.
 * Thus,  even the generated `.g.cpp` file has rvalue reference semantic, it is *not* going to be called as the caller does not have `move` semantic.
 * This makes the `const_cast` inevitable.